Add docker_login module

olsaki opened the pull request:
- Ansible version of "docker login" CLI command
- Persists Docker registry authentification in .dockercfg (only login once - no need to specify credentials over and over again anymore)
- Works for all other docker-py based modules (docker, docker_images) as well as the Docker CLI client

resmo commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,243 @@
+---
+module: docker_login
+author: Olaf Kilian
+version_added: "1.9"
```
should be 2.0

resmo commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,243 @@
+    description:
+       - URL of the registry, for example: https://index.docker.io/v1/
+    required: true
+    default: null
```
there is no default if required. can be removed.

resmo commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,243 @@
+       - URL of the registry, for example: https://index.docker.io/v1/
+    required: true
+    default: null
+    aliases: []
```
No aliases is the default. can be removed.

resmo commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,243 @@
+    from urlparse import urlparse
+except ImportError, e:
+    print "failed=True msg='failed to import python module: %s'" % e
+    sys.exit(1)
```
Do not use print and sys.exit in modules. Look at this example here  https://github.com/ansible/ansible-modules-extras/blob/devel/cloud/cloudstack/cs_affinitygroup.py#L95
and here https://github.com/ansible/ansible-modules-extras/blob/devel/cloud/cloudstack/cs_affinitygroup.py#L209
how you fail with module.fail_json if a python module is missing

resmo commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,243 @@
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            registry        = dict(required=True, default=None),
```
There is no default if `required=True`

resmo commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,243 @@
+
+        # Update the dockercfg if changed but not failed.
+        if self.has_changed():
+            self.update_dockercfg()
```
If I understand this right, It would be easy to implement check mode here if you add 

``` python
if self.has_changed() and not self.module.check_mode: 
     self.update_dockercfg()
```

and don't forget to add `supports_check_mode=True` just like this exmaple. https://github.com/ansible/ansible-modules-extras/blob/devel/cloud/cloudstack/cs_affinitygroup.py#L206

olsaki commented:
@resmo thanks for the review! Will update the PR asap.

olsaki commented:
@resmo thanks for your suggestions. I have updated the PR - can you have a look again please.

resmo commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,244 @@
+  email:
+    description:
+       - The email address for the registry account
+    required: false
```
there should be a default if not required

resmo commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,244 @@
+'''
+
+import os.path
+import sys
```
imho this is not used. can be renoved.

olsaki replied:
Removed "sys" import, but "os.path" is used here https://github.com/olsaki/ansible-modules-extras/blob/devel/cloud/docker/docker_login.py#L128 and here https://github.com/olsaki/ansible-modules-extras/blob/devel/cloud/docker/docker_login.py#L177

resmo commented:
2 minor things, otherwise looks good to me!

olsaki commented:
@resmo thanks again, good points, corrected those! Also added more meaningful error messages on registry login errors. I would appreciate a final review.

olsaki commented:
Anybody else? @bcoca is this clean to merge or any other fixes or rework needed?

gregdek commented:
ping @bcoca / @jimi-c / @abadger for mergey-ness

abadger commented:
Earlier I had wondered if this was the way we wanted authentication to work
(as separate modules) or if we wanted it to be integrated.  @bcoca
mentioned that there was precedent in another cloud module but we didn't
settle on whether it's a good architecture or not. We didn't discuss this
in the context of the new extras policies, though.  Is this some thing that
should be discussed amongst the current module authors on the mailing
list?  Our do we just want to have one current module author ok the
strategy?  Our something else?

bcoca commented:
I think it is a good question to raise in ansible-devel

olsaki commented:
Any news about this? Where can I follow the discussion?

gregdek commented:
@bcoca @jimi-c @abadger might I recommend that this be the topic of our next hangout?

gregdek commented:
@olsaki Sorry for the delay here, we'll get a conversation started about this soon.

abadger commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,246 @@
+  registry:
+    description:
+       - URL of the registry, for example: https://index.docker.io/v1/
+    required: true
```
We probably want this to be optional, defaulting to docker hub

abadger commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,246 @@
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            registry        = dict(required=True),
```
Same as above -- make this optional and default to docker hub.

abadger commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,246 @@
+    required: false
+    default: 600
+
+requirements: [ "docker-py" ]
```
Add a requirements "python >= 2.6"

Since docker-py requires newer python (that lets us know that we don't have to make this module compatible with pyhton-2.4.)

abadger commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,246 @@
+            registry        = dict(required=True),
+            username        = dict(required=True),
+            password        = dict(required=True),
+            email           = dict(required=False, default='anonymous@localhost.local'),
```
Is this a sensible default?  I don't know what really happens to the email address inside of docker-py so I'm not sure if we must pass this into the API at all?

abadger replied:
Ah -- from the commit history and comments, I see that you added this in response to the prior review.  I think that may have been a misunderstanding.  If the docker-py api is fine taking None here, we probably want the default to be None and document that the default is None in the documentation.

abadger commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,246 @@
+# import module snippets
+from ansible.module_utils.basic import *
+
+main()
```
Stylistically it's nice to use this:

``` python
if __name__ == '__main__':
    main()
```

That way if we ever need unittests for this module we can import it without running the module.

abadger commented:
Talked about this at today's community meeting and decided we're fine with the idea.  it is similar in function to openstack/os_auth.py.

I made some comments to fix up in the module then we can merge.

Is there anything about the docker modules in the core repository that should change because of this module?

gregdek commented:
@olsaki back to you. :)

memelet commented:
Would be nice if 'owner' and 'group' were arguments. When creating creds for a specific user (which I would guess is a common case) the .dockercfg file created is owned by root.

gregdek commented:
Adding new process. We will be evaluating all new module PRs according to this process, effective immediately.

# 

Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

Thanks again for submitting your Ansible module!

gregdek commented:
@olsaki a friendly ping to remind you that we're waiting on your changes based on @abadger comments.

olsaki commented:
pong - back on it, sorry for the delay.

olsaki commented:
@gregdek Made the requested changes. Please give me a day to test the module again before updating the branch.

olsaki commented:
@abadger from my point of view there is no change required in any core module. This module integrates with all other docker-py based modules (docker, docker_images) as well as the Docker CLI client, by simply persisting the required login credentials.

olsaki commented:
There seem to be some changes in the underlaying docker-py now since the initial submission of this module. I have already adapted to some of those (like new location and structure of the auth file) but need some more time to look into others (like handling of private registries). Working on it.

olsaki commented:
works_for_me

olsaki commented:
@memelet nice idea, from my perspective we easily can do that. But I would like to wait with that until this module got initially accepted.

abadger commented:
Reviewed changes and they look good.

Do we need to specify a minimum version of docker-py to work with this?  Some of the changes look like the API may be backwards incompatible.

abadger commented:
olaski - if you're on IRC.freenode.net, I'm abadger1999 in #ansible-devel.  Feel free to ping me if you want to.  This is very close to being done and I don't want it to fall off my radar by accident ;-)

olsaki commented:
@abadger thanks for the fast reply.

I had some issues with docker-py < 1.1.0 (auth not working properly).

The module has been successfully tested with docker-py 1.1.0 and from what I have seen should be compatible with all recent versions upwards from there. However, there can be incompatibility with older versions of the docker CLI client that expects the auth config to be in a different location and format.

Will give #ansible-devel a visit on the next chance!

abadger commented:
@olsaki Cool.  So we should probably list minimum versions if we know them in the documentation (the requirements: field.  It's free form but we try do something like ["libraryname", "library2 >= 1.1.0", "application >= 1.0"].  And we should also check the docker-py version at the same time as we check that we're able to import it.  I think that will be all.

olsaki commented:
@abadger ok, will add that info as advised. Just need some time to clarify the min. version of the docker cli tool to work with this and will get this back to you asap.

abadger commented:
@olsaki Cool thanks.  Note that 2.0 is going to go from beta to rc pretty soon.  We're going to make the cutoff for new modules then so it would be good to get this in quickly.  If that's going to be difficult to verify the minimum version, we could merge this with a version that's somewhat too stringent (the minimum version that you've tested with) and then relax it later when you find out more information.

(And please feel free to ping me on IRC if you update but I don't seem to notice here... release is always a hectic time and it's easy to miss something.)

abadger commented on cloud/docker/docker_login.py:
```diff
@@ -0,0 +1,251 @@
+        argument_spec = dict(
+            registry        = dict(required=False, default='https://index.docker.io/v1/'),
+            username        = dict(required=True),
+            password        = dict(required=True),
```
Noticed one more thing:  we should mark this as no_log like this:

``` python
    password    = dict(required=True, no_log=True),
```

That way ansible will apply some heuristics to keep that parameter from being displayed in output, logged, etc.

olsaki commented:
@abadger thanks for your suggestions and support. Much appreciated! Implemented it and back to you. Sorry for the delay, hope I'm not just too late with it.
